Add adr-002 evm hooks proposal

yihuang opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

ref impl: https://github.com/tharsis/ethermint/pull/417

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

yihuang commented on docs/architecture/adr-002-evm-hooks.md:
````diff
@@ -0,0 +1,40 @@
+}
+```
+
+- `PostTxProcessing` is called after EVM transaction finished, executed with the same cache context as the EVM
````
done, added the bank send example.

yihuang commented on docs/architecture/adr-002-evm-hooks.md:
````diff
@@ -0,0 +1,40 @@
+
+```golang
+type EvmHooks interface {
+	  PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error
````
yeah, it's possible, it's just I don't see an immediately use case, maybe someone could added in another proposal when actual use case appeared.

fedekunze replied:
the implementation in #417 doesn't receive the context. Why the difference?

yihuang replied:
> the implementation in #417 doesn't receive the context. Why the difference?

In newer version, it's the same as the adr.

fedekunze commented on docs/architecture/adr-002-evm-hooks.md:
````diff
@@ -0,0 +1,52 @@
+```
+
+- `PostTxProcessing` is called after EVM transaction finished, executed with the same cache context as the EVM
+  transaction execution, if `PostTxProcessing` returns an error, the whole EVM transaction is reverted.
````
explain why are you reverting the entire tx vs just doing a no-op for the hook post tx processing

yihuang replied:
added.
it makes the implementor to decide if certain api want to revert on error or not.

fedekunze commented on docs/architecture/adr-002-evm-hooks.md:
```diff
@@ -0,0 +1,52 @@
+
+## Abstract
+
+The current ADR proposes a hook interface to EVM module, with the goal of extend the functionality externally,
```
> the functionality

explain which functionality

yihuang replied:
changed to tx processing logic.

fedekunze commented on docs/architecture/adr-002-evm-hooks.md:
```diff
@@ -0,0 +1,52 @@
+## Context
+
+There are some requirements to enable EVM contract to communicate with cosmos native modules, which could have multiple
+use cases. We need some kind of hooks to support these extension requirements in a generic way.
```
explain use cases

yihuang replied:
Added more detail to the "Call Native Module" use case.

fedekunze commented on docs/architecture/adr-002-evm-hooks.md:
```diff
@@ -0,0 +1,52 @@
+The current ADR proposes a hook interface to EVM module, with the goal of extend the functionality externally,
+specifically to support communication from EVM contract to native modules through logs.
+
+## Context
```
This requires a more descriptive paragraph of the problem and the limitations of the current implementation.

> This section describes the forces at play, including technological, political, social, and project local. These forces are probably in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts. It should clearly explain the problem and motivation that the proposal aims to resolve.

yihuang replied:
Added some explanation for the motives to add hook interface.

fedekunze commented on docs/architecture/adr-002-evm-hooks.md:
````diff
@@ -0,0 +1,52 @@
+```golang
+evmKeeper.SetHooks(NewHook());
+```
+
````
There are a bunch of sections missing here:

- Backwards Compatibility
- Positive
- Negative 
- Neutral
- Further Discussions
- Test Cases
- References

yihuang replied:
added

fedekunze reviewed (changes requested):
The ADR is good, but it doesn't follow the ADR template and a lot of context and details are missing if you compare it with the implementation. For example, the following decision is not documented on the ADR:

```
if !res.Failed() {
	err = k.PostTxProcessing(txHash, logs)
	if err != nil {
		k.RevertToSnapshot(revision)
		res.VmError = "failed to process native logs"
		res.Ret = []byte(err.Error())
	}
}
```

Some questions that need to be documented:

- why do we revert the state to the snapshot prior to the tx execution? (i.e why did you choose this approach over just performing a no-op on `PostTxProcessing`)
- give a reference implementation for a `Hook` implementation
- explain the `MultiHook` design for adding multiple hooks
- explain how to add the Hook to the `app.go`

yihuang commented:
> The ADR is good, but it doesn't follow the ADR template and a lot of context and details are missing if you compare it with the implementation. For example, the following decision is not documented on the ADR:
> 
> ```
> if !res.Failed() {
> 	err = k.PostTxProcessing(txHash, logs)
> 	if err != nil {
> 		k.RevertToSnapshot(revision)
> 		res.VmError = "failed to process native logs"
> 		res.Ret = []byte(err.Error())
> 	}
> }
> ```

added a little bit more explanation to this:
- hooks are called only when tx executed successfully
- errors returned by hooks get translated to vm error, and cause the whole tx to revert.

> 
> Some questions that need to be documented:
> 
> * why do we revert the state to the snapshot prior to the tx execution? (i.e why did you choose this approach over just performing a no-op on `PostTxProcessing`)

added, with this approach, the implementor is still free to swallow any error.

> * give a reference implementation for a `Hook` implementation

add the `BankSendHook` to adr.

> * explain the `MultiHook` design for adding multiple hooks

I think this is not essential for the proposal?

> * explain how to add the Hook to the `app.go`

added.

fedekunze reviewed (commented):
Looks better now, although there are some details and code snippet are missing about the integration with the `state_transition.go`. 

I will also recommend you use [Grammarly](https://www.grammarly.com/) to fix some of the wording.

yihuang commented on docs/architecture/adr-002-evm-hooks.md:
```diff
@@ -0,0 +1,175 @@
+
+func (h BankSendHook) PostTxProcessing(ctx sdk.Context, txHash ethcmn.Hash, logs []*ethtypes.Log) error {
+	for _, log := range logs {
+		if len(log.Topics) > 0 && log.Topics[0] == BankSendEvent.ID {
```
You mean whitelisting contracts that are allowed to do this call? that may be a business decision to make, as long as the call is secure, I don't see why we shouldn't open it to all contracts.

fedekunze replied:
> that may be a business decision to make, as long as the call is secure, I don't see why we shouldn't open it to all contracts.

for example: what if someone creates an event with the same name that doesn't deduct the funds from the user? 

In general, I think it's ok to leave the assumption that the application will be in charge of the security of the execution of state changes with regards to using logs in conjunction with SDK modules, although we should definitely explain that that on the `Neutral` consequences with a warning.

fedekunze replied:
also add the comment on the ADR above the `if len(log.Topics) == 0 || log.Topics[0] != BankSendEvent.ID` condition

yihuang replied:
Yeah, I agree that the potential name conflict should be put into `Neutral` or even `Negative` consequence.
I'll add an item into `Negative` consequence, and add comments in the code.

yihuang replied:
done

yihuang commented on docs/architecture/adr-002-evm-hooks.md:
```diff
@@ -0,0 +1,175 @@
+
+<!-- Test cases for an implementation are mandatory for ADRs that are affecting consensus changes. Other ADRs can choose to include links to test cases if applicable. -->
+
+## References
```
done

yihuang commented on docs/architecture/adr-002-evm-hooks.md:
```diff
@@ -0,0 +1,175 @@
+			unpacked, err := BankSendEvent.Inputs.Unpack(log.Data)
+			if err != nil {
+				// ignore unrecognizable log
+				continue
```
done

yihuang commented:
> Looks better now, although there are some details and code snippet are missing about the integration with the `state_transition.go`.

done, added some `ApplyTransaction` code snippet.

> 
> I will also recommend you use [Grammarly](https://www.grammarly.com/) to fix some of the wording.

done

fedekunze reviewed (approved):
ACK 👍
